### PR TITLE
fix: AwsSolutions-APIG4 triggers for CORS preflight endpoints

### LINKED ISSUE #1815: bug: AwsSolutions-APIG4 triggers for CORS preflight endpoints
### What is the problem?

I am using AWS CDK to create an API Gateway with default CORS preflight endpoints. This then spins up an OPTIONS endpoint for each resource I have in my Rest API, which does not have authorization. This causes the following error from `cdk-nag`:
```
[Error at /Stack/TestRestApi/Default/test/OPTIONS/Resource] AwsSolutions-APIG4: The API does not implement authorization. In most cases an API needs to have an authentication and authorization implementation strategy. This includes using such approaches as IAM, Cognito User Pools, Custom authorizer, etc.
```

Looking at the [fetch specification](https://fetch.spec.whatwg.org/#cors-protocol-and-credentials), preflight requests are not sent with authorization by clients, therefore we shouldn't enable authorization on those endpoints:
> a CORS-preflight request never includes credentials.

`cdk-nag` should therefore not error `AwsSolutions-APIG4` on CORS preflight endpoints.

### Reproduction Steps

Example Rest API in CDK with default CORS preflight enabled:
```typescript
import apigateway from 'aws-cdk-lib/aws-apigateway';

new apigateway.RestApi(this, 'TestRestApi', {
    restApiName: 'Test',
    defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
    },
});
```

### What did you expect to happen?

No error for `AwsSolutions-APIG4`.

### What actually happened?

Error for `AwsSolutions-APIG4`.

### cdk-nag version

2.29.12

### Language

Typescript

### Other information

Possibly related to #1100.

### PR DESCRIPTION
**Pull request by georeeve**
Fixes #1815.

Currently this PR just marks all `OPTIONS` methods as compliant. I'm not sure the best way to narrow this down to specifically CORS preflight endpoints. We could try and match the template that CDK generates through its `addCorsPreflight` method, but that won't include any custom preflight endpoints which people may be using, and any changes to the CDK template may break our checks here.

I took a look at the [MDN docs for the `OPTIONS` method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS), and it seems like it has limited use cases outside of CORS. Perhaps we can mark them all as compliant?

Here is the CloudFormation CfnMethod template which CDK generates when using the following code:
```typescript
new apigateway.RestApi(this, 'TestRestApi', {
    restApiName: 'Test',
    defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
    },
});
```

```json
{
  "apiKeyRequired": false,
  "authorizationType": "NONE",
  "httpMethod": "OPTIONS",
  "integration": {
    "type": "MOCK",
    "requestTemplates": {
      "application/json": "{ statusCode: 200 }"
    },
    "integrationResponses": [
      {
        "statusCode": "204",
        "responseParameters": {
          "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
          "method.response.header.Access-Control-Allow-Origin": "'*'",
          "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'"
        }
      }
    ]
  },
  "methodResponses": [
    {
      "statusCode": "204",
      "responseParameters": {
        "method.response.header.Access-Control-Allow-Headers": true,
        "method.response.header.Access-Control-Allow-Origin": true,
        "method.response.header.Access-Control-Allow-Methods": true
      }
    }
  ],
  "resourceId": {
    "Fn::GetAtt": ["Test", "RootResourceId"]
  },
  "restApiId": {
    "Ref": "Test"
  }
}
```

### CODE REVIEW

**Comment by dontirun** [comment]
I thinks it's better to keep the current implementation of the rule and add additional information on the description that it may not apply to pre-flight endpoints Instead of marking all `OPTIONS`as compliant.

While not ideal I'd rather users suppress false positives instead of opening up the rule for false negatives

**Comment by georeeve** [comment]
If we are concerned about false negatives, then we could at least take the other approach of matching the built-in CORS feature by matching against the `MOCK` integration outlined above. Again, we might have to change if it CDK changes their implementation, but I think it's better than simply updating the rule description and having users craft their own exemption config for what is a built-in feature to CDK.

**Comment by georeeve** [comment]
Perhaps we could be a bit more generic and check the above `methodResponses`? I don't imagine they're likely to change, and I can't see why you would return that on any other `OPTIONS` endpoint which is not dedicated to responding to CORS preflight requests.

**Comment by georeeve** [comment]
I've just updated the check to match the `methodResponses` responses above. I think this is a better approach than just updating the description, but please have a look and let me know.

**Comment by dontirun** [comment]
> However, I did try and come up with an exemption config which only exempts `OPTIONS` methods, but I couldn't get it to work for some reason. Do you have any examples that I could use if so?

You can mark that specific case as [NagRuleCompliance.NOT_APPLICABLE](https://github.com/cdklabs/cdk-nag/blob/c40f7255a9ff6dab71d14e4188047065036b903e/docs/RuleCreation.md#anatomy-of-a-rule)

File: src/rules/apigw/APIGWAuthorization.ts [hunk 1]
```diff
@@ -3,18 +3,41 @@ Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
...
  * APIs implement authorization
  * @param node the CfnResource to check
  */
 export default Object.defineProperty(
   (node: CfnResource): NagRuleCompliance => {
     if (node instanceof CfnMethod || node instanceof CfnRoute) {
+      if (node instanceof CfnMethod) {
+        const httpMethod = NagRules.resolveIfPrimitive(node, node.httpMethod);
+        if (httpMethod === 'OPTIONS' && checkCORSMethodResponses(node)) {
+          return NagRuleCompliance.COMPLIANT;
```
Comment (dontirun) [inline src/rules/apigw/APIGWAuthorization.ts:38]: Given the spec I think it's safe to change this from `COMPLIANT` to `NOT_APPLICABLE`

**Review by dontirun — CHANGES_REQUESTED** [review]

File: test/rules/APIGW.test.ts [hunk 2]
```diff
@@ -193,7 +194,11 @@ describe('Amazon API Gateway', () => {
       });
       validateStack(stack, ruleId, TestType.NON_COMPLIANCE);
     });
-    test('Compliance', () => {
+    test('Noncompliance 3', () => {
+      new RestApi(stack, 'rRestApi').root.addMethod('OPTIONS');
```
[the remark this answers is not in the document]
Comment (georeeve) [reply test/rules/APIGW.test.ts]: Done, I've just removed it from the tests within the describe block, let me know if you want it removed for the whole file.

**Review by dontirun — APPROVED** [review]

**Pull request merged**